primitives/arithmetic/src/lib.rs

```diff
@@ -40,13 +40,96 @@ mod per_things;
 mod fixed;
 mod rational128;
 
-pub use fixed::{FixedPointNumber, Fixed64, Fixed128};
-pub use per_things::{PerThing, Percent, PerU16, Permill, Perbill, Perquintill};
+pub use fixed::{FixedPointNumber, FixedPointOperand, Fixed64, Fixed128};
+pub use per_things::{PerThing, InnerOf, Percent, PerU16, Permill, Perbill, Perquintill};
 pub use rational128::Rational128;
 
+use sp_std::cmp::Ordering;
+use sp_std::ops;
+use traits::{UniqueSaturatedInto, Saturating};
+
+/// Trait for comparing two numbers with an epsilon coefficient.
+///
+/// Returns:
+/// - `Ordering::Greater` if `self` is greater than `other + other * epsilon`.
+/// - `Ordering::Less` if `self` is less than `other - other * epsilon`.
+/// - `Ordering::Equal` otherwise.
+pub trait EpsilonOrd<T, P> {
+	/// Compare if `self` is `epsilon` greater or less than `other`.
+	fn ecmp(&self, other: &T, epsilon: P) -> Ordering;
+}
+
+impl<T, P> EpsilonOrd<T, P> for T
+where
+	P: PerThing,
+	T: Ord + PartialOrd + Clone + Copy + ops::Rem<T, Output=T> + ops::Mul<T, Output=T> +
+		ops::Div<T, Output=T> + ops::Add<T, Output=T> + UniqueSaturatedInto<InnerOf<P>> +
+		From<InnerOf<P>> + Saturating,
+{
+	fn ecmp(&self, other: &T, epsilon: P) -> Ordering {
```

Comment:
I find `epsilon: Perthing` a bit strange.  I would have expected `epsilon: T` or something else. It seems that `ecmp` violates commutativity (or was it called asymmetry?):
```rust
        let f = Perbill::from_percent(100);
		assert_eq!(50u32.ecmp(&5u32, f), Ordering::Greater);
		assert_eq!(5u32.ecmp(&50u32, f), Ordering::Equal);
```
 I would have expected `Greater` and `Less`, or both `Equal`.

Reply:
I somewhat agree, I first started with with a generic that can be `Fixed` but that seemed like an overkill. 

Most often when you use the term _epsilon_ you mean a value the limit of which at infinity is zero, i.e. _a very small value_. PerThing should be fine for this use case. 

The asymmetry situation needs a fix.

Reply:
Maybe define it as follows?
```rust
let (diff, negative) = if self > other { (self - other, false) } else { (other - self, true) };
let coeff = Perbill::from_rational_approximation(diff, if negative { self } else { other } // or choose smaller one);
if coeff > epsilon {
 if negative { Ordering::Greater } else { Ordering::Less } // or the other way around :thinking: 
} else {
  Ordering::Equal
}
```

Reply:
After some thinking, my current decision is to keep it as is. Here's why: 

1. No this is not [commutative](https://en.wikipedia.org/wiki/Commutative_property) by definition. What we want to know is: given `ecmp(a, b, c)`, is `a` bigger than `b+c`? is it smaller than `b-c`, or is it equal to `b` with an error rate of `c`. 

Of course, if you change the operands, the result will differ. But I don't see this as a problem at all. Heck, normal `Ord` is also never commutative. `a.cmp(b)` is never the same as `b.cmp(a)`. Except, in normal `Ord` you can assume that `a.cmp(b)` is always `!b.cmp(a)` but here you cannot assume this. 

@tomusdrw's implementation addresses this by always comparing with the smaller value, which I think is not needed based on the above argument. 

Raise a hand if you disagree.

Reply:
I don't have a strong opinion, I was just pointing it out 👍.

Reply:
In the example you used an absolute epsilon (`-c / +c`) and then it makes perfect sense - the relation is symmetric. Using `Perbill` however makes this feel a bit weird that the arguments order matters and may lead to different results.

For instance `99.ecmp(110, 10%) == Equal` but `110.ecmp(99, 10%) == Greater`.

In practice with Phragmén the numbers we compare are going to be quite similar and the `epsilon` values are going to be really small, but since the code is placed in a more generic place it might lead to some confusion in the future.

Reply:
sure, then I'll call it `ThresholdOrd` since I think epsilon only makes sense for values less than 1.